Add support for explicit broadcasting in TensorFlow

yongtang opened the pull request:
This fix tries to adds support for explicit broadcasting in TensorFlow, as was suggested in #14509. This fix adds the op of tf.broadcast_to, which is equivalent to the numpy.broadcast_to in numpy.

This fix fixes #14509.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add support for explicit broadcasting in TensorFlow

This fix tries to adds support for explicit broadcasting in TensorFlow,
as was suggested in 14509. This fix adds the op of tf.broadcast_to,
which is equivalent to the numpy.broadcast_to in numpy.

This fix fixes 14509.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Register BroadcastTo op in array_ops.cc

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Sanitize with clang-format -i

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases for tf.broadcast_to

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Sanitize bazel BUILD and python.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

caisq commented:
@rmlarsen Do you want to review this CL? @martinwicke I assume this requires an API review.

alextp commented on tensorflow/core/ops/array_ops.cc:
```diff
@@ -410,6 +410,20 @@ axis: Dimension along which to unpack.  Negative values wrap around, so the
 
+REGISTER_OP("BroadcastTo")
+    .Input("input: T")
+    .Input("shape: int32")
```
Might be worth it to support int64 shapes as well

yongtang replied:
Thanks @alextp. Done.

alextp commented on tensorflow/core/ops/array_ops.cc:
```diff
@@ -410,6 +410,20 @@ axis: Dimension along which to unpack.  Negative values wrap around, so the
+    .Input("shape: int32")
+    .Output("output: T")
+    .Attr("T: type")
+    .SetShapeFn([](InferenceContext* c) { return Status::OK(); })
```
You can do much more than a pass-through shape inference function here.

Specifically, if the shape tensor has a known constant value (and it often will) you should be able to set the output shape to be exactly that shape. See the shape_fn for reshape; you might be able to reuse most of its code here.

yongtang replied:
Done. Thanks.

alextp commented on tensorflow/core/kernels/broadcast_to_op.cc:
```diff
@@ -0,0 +1,222 @@
+        BROADCAST_SHAPE(broadcast, reshape, 1, input_shape, output_shape);
+
+        auto output = output_tensor->tensor<T, 1>();
+        switch (input_shape.dims()) {
```
Not allowing broadcasting from a scalar?

yongtang replied:
Thanks @alextp. The PR has been updated to processing scalar with `setConstant` to broadcast.

alextp commented on tensorflow/python/kernel_tests/array_ops_test.py:
```diff
@@ -1123,6 +1123,40 @@ def testResourceRejection(self):
                                                "cannot be a resource variable"):
         guarantee_a.eval()
 
+class BroadcastToTest(test_util.TensorFlowTestCase):
```
I think it'd be nice to have these tests in their own file, as this file is already quite long.

yongtang replied:
Done. Thanks.

alextp reviewed (changes requested):
Overall looks good, some minor comments

yongtang committed:
Split broadcast_to_ops_test from array_ops_test

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Support int64 shape

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Improve shape inference for broadcast_to

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add scalar input support for broadcast_to

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@alextp Thanks for the review. The PR has been updated. Please take a look and let me know if there are any issues.

alextp commented on tensorflow/core/kernels/broadcast_to_op.cc:
```diff
@@ -0,0 +1,239 @@
+
+// As MakeShape is able to handle both DT_INT32 and DT_INT64,
+// no need to have TypeConstraint for `Tidx`
+#define REGISTER_KERNEL(type)                                           \
```
Register a GPU kernel as well? That might require adding a .device() before the eigen call to .broadcast but otherwise should be easy.

alextp commented on tensorflow/core/ops/array_ops.cc:
```diff
@@ -410,6 +410,65 @@ axis: Dimension along which to unpack.  Negative values wrap around, so the
+
+input: A Tensor to broadcast.
+shape: An 1-D `int` Tensor. The shape of the desired output.
+output: A Tensor.
```
It might be useful to explain here how broadcasting works (align dimensions, 1s are expanded to be the size, etc)

alextp commented on tensorflow/python/kernel_tests/broadcast_to_ops_test.py:
```diff
@@ -0,0 +1,86 @@
+
+  def testBroadcastToBasic(self):
+    for dtype in [np.uint8, np.uint16, np.int8, np.int16, np.int32, np.int64]:
+      with self.test_session():
```
You want self.test_session(use_gpu=True) to make sure the GPU gets used when it's available (it's a no-op when the GPU is not available)

alextp reviewed (changes requested):
Thanks for the changes!

shoyer commented on tensorflow/python/kernel_tests/broadcast_to_ops_test.py:
```diff
@@ -0,0 +1,86 @@
+      v_tf = array_ops.broadcast_to(constant_op.constant(x), [3, 3])
+      v_np = np.broadcast_to(x, [3, 3])
+      self.assertAllEqual(v_tf.eval(), v_np)
+
```
Please take a look at NumPy's tests for `broadcast_to` as an example of what good coverage for shapes looks like:
https://github.com/numpy/numpy/blob/4f4b0a43c2c5c1441a507a214045a50a1bfc03ed/numpy/lib/tests/test_stride_tricks.py#L222-L265

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

martinwicke commented:
@av8ramit Is this nag (wording) appropriate since this has "awaiting response"?

av8ramit commented:
@martinwicke yeah might need to toggle the state machine a bit.

martinwicke commented:
@yongtang can you look at the latest comments?

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

yongtang committed:
Update API defs tensorflow/core/api_def/update_api_def.sh

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update API golden

```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
           --update_goldens True
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update docstring for broadcast_to

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
The doctoring has been updated. Will find a GPU machine to test and enable GPU kernel.

yongtang committed:
Enable GPU kernel for BroadcastTo

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Enable use_gpu=True for test cases

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

alextp commented on tensorflow/core/kernels/BUILD:
```diff
@@ -606,6 +606,7 @@ cc_library(
         ":batch_space_ops",
         ":bcast_ops",
         ":bitcast_op",
+        ":broadcast_to_op",
```
Why is this here?

yongtang replied:
@alextp The dependency is added so that it is part of the `gen_array_ops.py`.

alextp commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -28,6 +28,7 @@
 @@saturate_cast
 @@broadcast_dynamic_shape
 @@broadcast_static_shape
+@@broadcast_to
```
Let's add this to contrib for one release before we commit to supporting this API forever. To do so, remove this @@ annotation here, revert the change to tensorflow.pbtxt, and add visibility: HIDDEN to the apidef.pbtxt proto. Then, import this symbol (broadcast_to) from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/framework/__init__.py so it gets exposed as tf.contrib.framework.broadcast_to

martinwicke replied:
Can't you now simply do tf_export('tf.contrib.framework.broadcast_to')?

@annarev?

yongtang replied:
@alextp @martinwicke I add `tf_export('contrib.framework.broadcast_to')(broadcast_to)` and that seems to do the work. Although I also need to update `_allowed_symbols` in `tensorflow/contrib/framework/__init__.py` so that the symbol could be exposed.

yongtang replied:
Thanks @annarev. I updated the PR so that it is exposed from `tensorflow/contrib/framework/__init__.py`.

yongtang committed:
Hiden the ops and export to tf.contrib.framework for now.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add the op to the _allowed_symbol in tf.contrib.framework

Otherwise the symbole will be hidden

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix pylint sanity issue.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

jhseu committed:
Merge branch 'master' into 14509-broadcast_to

hadim commented:
Is that normal I can't see `tf.broadcast_to` in the last 1.8.0 version?

Hoeze commented:
I just compiled the "r1.9" branch, but there is still no "tf.broadcast_to":
```
>>> tf.__version__
'1.9.0-rc0'
>>> tf.broadcast_to
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'tensorflow' has no attribute 'broadcast_to'
```
How can I get a Tensorflow-Version supporting this?

martinwicke commented:
This PR added the op, but hid it by adding this file: tensorflow/core/api_def/python_api/api_def_BroadcastTo.pbtxt

Not sure whether that was intentional, but removing that file should make this op visible.

Hoeze commented:
Removing means deleting it?

martinwicke commented:
I think that would be sufficient.​

Hoeze commented:
Thank you very much, deleting that file and recompiling everything did the trick!

yongtang commented:
@Hoeze The op is exposed as `tf.contrib.framework.broadcast_to` for now (See https://github.com/tensorflow/tensorflow/pull/15243#discussion_r178889285):
```
# python
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> tf.__version__
'1.9.0-dev20180604'
>>> tf.contrib.framework.broadcast_to
<function broadcast_to at 0x7f1f8588c230>
>>> 
```

martinwicke commented:
​Could you send a PR? I don't see a reason to hide this op.

yongtang commented:
@martinwicke @Hoeze The PR #19753 has been created to expose `tf.broadcast_to`.

Hoeze commented:
@yongtang This patch should be merged to `r1.9` branch as well, shouldn't it?
